File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/features/livelist/view/singlecolumn/SingleColumnListView.java
package com.trtc.uikit.livekit.features.livelist.view.singlecolumn;

import android.content.Context;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleObserver;
import androidx.lifecycle.Observer;
import androidx.lifecycle.OnLifecycleEvent;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.tencent.cloud.tuikit.engine.extension.TUILiveListManager;
import com.trtc.uikit.livekit.R;
import com.trtc.uikit.livekit.common.LiveKitLogger;
import com.trtc.uikit.livekit.component.pictureinpicture.PictureInPictureStore;
import com.trtc.uikit.livekit.features.livelist.LiveListViewDefine;
import com.trtc.uikit.livekit.features.livelist.manager.LiveInfoListService;
import com.trtc.uikit.livekit.features.livelist.view.liveListviewpager.LiveListViewPager;
import com.trtc.uikit.livekit.features.livelist.view.liveListviewpager.LiveListViewPagerAdapter;

import java.util.HashSet;

public class SingleColumnListView extends FrameLayout {
    private static final LiveKitLogger LOGGER = LiveKitLogger.getComponentLogger("SingleColumnListView");

    private static final int REFRESH_TIME_INTERVAL = 1000;

    private final HashSet<SingleColumnItemView> mPlayStreamView = new HashSet<>(2);

    private FragmentActivity                       mFragmentActivity;
    private LiveListViewDefine.LiveListViewAdapter mLiveListViewAdapter;
    private LiveInfoListService                    mLiveInfoListService;

    private LiveListViewPager                      mLiveListViewPager;
    private LiveListViewPagerAdapter               mLiveListViewPagerAdapter;
    private SwipeRefreshLayout                     mSwipeRefreshLayout;
    private LiveListViewDefine.OnItemClickListener mOnItemClickListener;
    private SingleColumnItemView                   mWillEnterRoomView;
    private boolean                                mIsLoading;
    private boolean                                mIsResumed;
    private long                                   mLoadingTime;
    private boolean                                mIsInit;

    private final Observer<String> mPictureInPictureRoomIdObserver = this::onPictureInPictureRoomIdChanged;

    public SingleColumnListView(@NonNull Context context) {
        this(context, null, 0);
    }

    public SingleColumnListView(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public SingleColumnListView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        LayoutInflater.from(getContext()).inflate(R.layout.livelist_single_column_list_view, this, true);
        initView();
    }

    private void initView() {
        mSwipeRefreshLayout = findViewById(R.id.swipe_layout);
        mSwipeRefreshLayout.setColorSchemeResources(R.color.common_design_standard_g5);
        mLiveListViewPager = findViewById(R.id.live_list_view_pager);
        mSwipeRefreshLayout.setOnRefreshListener(this::refreshData);
    }

    public void init(@NonNull FragmentActivity fragmentActivity, LiveListViewDefine.LiveListViewAdapter adapter,
                     LiveInfoListService liveInfoListService) {
        mFragmentActivity = fragmentActivity;
        mLiveListViewAdapter = adapter;
        mLiveInfoListService = liveInfoListService;
        mPlayStreamView.clear();
        mLiveListViewPagerAdapter = new LiveListViewPagerAdapter(mFragmentActivity,
                mLiveInfoListService) {
            @Override
            public View createLiveInfoView(TUILiveListManager.LiveInfo liveInfo) {
                SingleColumnItemView singleColumnItemView = new SingleColumnItemView(getContext());
                singleColumnItemView.createLiveInfoView(mLiveListViewAdapter, liveInfo);
                singleColumnItemView.setOnClickListener(view -> {
                    mWillEnterRoomView = (SingleColumnItemView) view;
                    if (mOnItemClickListener != null) {
                        mOnItemClickListener.onItemClick(view, liveInfo);
                    }
                });
                return singleColumnItemView;
            }

            @Override
            public void updateLiveInfoView(View view, TUILiveListManager.LiveInfo liveInfo) {
                SingleColumnItemView singleColumnItemView = (SingleColumnItemView) view;
                singleColumnItemView.updateLiveInfoView(liveInfo);
                singleColumnItemView.setOnClickListener(v -> {
                    mWillEnterRoomView = (SingleColumnItemView) v;
                    if (mOnItemClickListener != null) {
                        mOnItemClickListener.onItemClick(view, liveInfo);
                    }
                });
            }

            @Override
            public void onViewDidSlideIn(View view) {
                SingleColumnItemView itemView = (SingleColumnItemView) view;
                startPreviewLiveStream(itemView, false);
            }

            @Override
            public void onViewWillSlideIn(View view) {
                SingleColumnItemView itemView = (SingleColumnItemView) view;
                startPreviewLiveStream(itemView, true);
            }

            @Override
            public void onViewDidSlideOut(View view) {
                SingleColumnItemView itemView = (SingleColumnItemView) view;
                stopPreviewLiveStream(itemView);
            }

            @Override
            public void onViewSlideInCancelled(View view) {
                SingleColumnItemView itemView = (SingleColumnItemView) view;
                stopPreviewLiveStream(itemView);
            }
        };
        mLiveListViewPager.setAdapter(mLiveListViewPagerAdapter);
        mLiveListViewPagerAdapter.fetchData();
    }

    public void setOnItemClickListener(LiveListViewDefine.OnItemClickListener listener) {
        mOnItemClickListener = listener;
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        if (mFragmentActivity != null) {
            mFragmentActivity.getLifecycle().addObserver(mLifecycleObserver);
        }
        PictureInPictureStore.sharedInstance().getState().roomId.observeForever(mPictureInPictureRoomIdObserver);
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        if (mFragmentActivity != null) {
            mFragmentActivity.getLifecycle().removeObserver(mLifecycleObserver);
        }
        stopAllPreviewLiveStream();
        PictureInPictureStore.sharedInstance().getState().roomId.removeObserver(mPictureInPictureRoomIdObserver);
    }

    private void startPreviewLiveStream(SingleColumnItemView itemView, boolean isMuteAudio) {
        itemView.startPreviewLiveStream(isMuteAudio);
        mPlayStreamView.add(itemView);
    }

    private void stopPreviewLiveStream(SingleColumnItemView itemView) {
        itemView.stopPreviewLiveStream();
        mPlayStreamView.remove(itemView);
    }

    private void stopAllPreviewLiveStream() {
        for (SingleColumnItemView itemView : mPlayStreamView) {
            itemView.stopPreviewLiveStream();
        }
        mPlayStreamView.clear();
    }

    private void pauseAllLiveStream() {
        for (SingleColumnItemView itemView : mPlayStreamView) {
            if (mWillEnterRoomView == itemView) {
                continue;
            }
            itemView.stopPreviewLiveStream();
        }
    }

    public void refreshData() {
        if (mIsLoading) {
            return;
        }

        if (System.currentTimeMillis() - mLoadingTime < REFRESH_TIME_INTERVAL) {
            removeCallbacks(mRunnable);
            postDelayed(mRunnable, REFRESH_TIME_INTERVAL);
            return;
        }

        mLoadingTime = System.currentTimeMillis();
        mIsLoading = true;
        if (mLiveListViewPagerAdapter != null) {
            mLiveListViewPagerAdapter.refreshData(() -> {
                mSwipeRefreshLayout.setRefreshing(false);
                post(() -> {
                    mIsLoading = false;
                    if (!mIsResumed) {
                        LOGGER.info("the current activity is not resumed");
                        return;
                    }
                    stopAllPreviewLiveStream();
                    int currentPosition = mLiveListViewPager.getCurrentItem();
                    View currentView = mLiveListViewPager.findViewByPosition(currentPosition);
                    if (currentView != null) {
                        SingleColumnItemView columnItemView = (SingleColumnItemView) currentView;
                        startPreviewLiveStream(columnItemView, false);
                    }
                });
            });
        }
    }

    private void onPictureInPictureRoomIdChanged(String roomId) {
        if (TextUtils.isEmpty(roomId)) {
            for (SingleColumnItemView itemView : mPlayStreamView) {
                if (itemView.isPauseByPictureInPicture()) {
                    itemView.startPreviewLiveStream(false);
                }
            }
        }
    }

    private final LifecycleObserver mLifecycleObserver = new LifecycleObserver() {
        @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
        public void onResume() {
            mIsResumed = true;
            mWillEnterRoomView = null;
            if (mIsInit) {
                refreshData();
            } else {
                mIsInit = true;
            }
        }

        @OnLifecycleEvent(Lifecycle.Event.ON_PAUSE)
        public void onPause() {
            mIsResumed = false;
            pauseAllLiveStream();
        }
    };

    private final Runnable mRunnable = new Runnable() {
        @Override
        public void run() {
            if (!mIsLoading) {
                mSwipeRefreshLayout.setRefreshing(false);
            }
        }
    };
}

File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/features/livelist/view/singlecolumn/SingleColumnListView.kt
package com.trtc.uikit.livekit.features.livelist.view.singlecolumn

import android.content.Context
import android.text.TextUtils
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.FrameLayout
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleEventObserver
import androidx.lifecycle.Observer
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.tencent.cloud.tuikit.engine.extension.TUILiveListManager
import com.trtc.uikit.livekit.R
import com.trtc.uikit.livekit.common.LiveKitLogger
import com.trtc.uikit.livekit.component.pictureinpicture.PictureInPictureStore
import com.trtc.uikit.livekit.features.livelist.LiveListViewAdapter
import com.trtc.uikit.livekit.features.livelist.OnItemClickListener
import com.trtc.uikit.livekit.features.livelist.manager.LiveInfoListService
import com.trtc.uikit.livekit.features.livelist.view.liveListviewpager.LiveListViewPager
import com.trtc.uikit.livekit.features.livelist.view.liveListviewpager.LiveListViewPagerAdapter

class SingleColumnListView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : FrameLayout(context, attrs, defStyleAttr) {

    companion object {
        private val LOGGER = LiveKitLogger.getComponentLogger("SingleColumnListView")
        private const val REFRESH_TIME_INTERVAL = 1000L
    }

    private lateinit var fragmentActivity: FragmentActivity
    private lateinit var liveListViewAdapter: LiveListViewAdapter
    private lateinit var liveInfoListService: LiveInfoListService
    private lateinit var liveListViewPager: LiveListViewPager
    private lateinit var swipeRefreshLayout: SwipeRefreshLayout
    private lateinit var liveListViewPagerAdapter: LiveListViewPagerAdapter

    private val playStreamView = HashSet<SingleColumnItemView>(2)
    private var onItemClickListener: OnItemClickListener? = null
    private var willEnterRoomView: SingleColumnItemView? = null
    private var isLoading = false
    private var isResumed = false
    private var loadingTime = 0L
    private var isInit = false

    private val pictureInPictureRoomIdObserver = Observer<String> { roomId ->
        onPictureInPictureRoomIdChanged(roomId)
    }

    init {
        LayoutInflater.from(context).inflate(R.layout.livelist_single_column_list_view, this, true)
        initView()
    }

    private fun initView() {
        swipeRefreshLayout = findViewById(R.id.swipe_layout)
        swipeRefreshLayout.setColorSchemeResources(R.color.common_design_standard_g5)
        liveListViewPager = findViewById(R.id.live_list_view_pager)
        swipeRefreshLayout.setOnRefreshListener { refreshData() }
    }

    fun init(
        fragmentActivity: FragmentActivity,
        adapter: LiveListViewAdapter,
        liveInfoListService: LiveInfoListService
    ) {
        this.fragmentActivity = fragmentActivity
        this.liveListViewAdapter = adapter
        this.liveInfoListService = liveInfoListService
        playStreamView.clear()

        liveListViewPagerAdapter = object : LiveListViewPagerAdapter(fragmentActivity, liveInfoListService) {
            override fun createLiveInfoView(liveInfo: TUILiveListManager.LiveInfo): View {
                return SingleColumnItemView(context).apply {
                    createLiveInfoView(liveListViewAdapter, liveInfo)
                    setOnClickListener { view ->
                        willEnterRoomView = view as SingleColumnItemView
                        onItemClickListener?.onItemClick(view, liveInfo)
                    }
                }
            }

            override fun updateLiveInfoView(view: View, liveInfo: TUILiveListManager.LiveInfo) {
                (view as SingleColumnItemView).apply {
                    updateLiveInfoView(liveInfo)
                    setOnClickListener { v ->
                        willEnterRoomView = v as SingleColumnItemView
                        onItemClickListener?.onItemClick(v, liveInfo)
                    }
                }
            }

            override fun onViewDidSlideIn(view: View) {
                startPreviewLiveStream(view as SingleColumnItemView, false)
            }

            override fun onViewWillSlideIn(view: View) {
                startPreviewLiveStream(view as SingleColumnItemView, true)
            }

            override fun onViewDidSlideOut(view: View) {
                stopPreviewLiveStream(view as SingleColumnItemView)
            }

            override fun onViewSlideInCancelled(view: View) {
                stopPreviewLiveStream(view as SingleColumnItemView)
            }
        }

        liveListViewPager.setAdapter(liveListViewPagerAdapter)
        liveListViewPagerAdapter.fetchData()
    }

    fun setOnItemClickListener(listener: OnItemClickListener) {
        onItemClickListener = listener
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        fragmentActivity.lifecycle.addObserver(lifecycleObserver)
        PictureInPictureStore.sharedInstance().state.roomId.observeForever(pictureInPictureRoomIdObserver)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        fragmentActivity.lifecycle.removeObserver(lifecycleObserver)
        stopAllPreviewLiveStream()
        PictureInPictureStore.sharedInstance().state.roomId.removeObserver(pictureInPictureRoomIdObserver)
    }

    private fun startPreviewLiveStream(itemView: SingleColumnItemView, isMuteAudio: Boolean) {
        itemView.startPreviewLiveStream(isMuteAudio)
        playStreamView.add(itemView)
    }

    private fun stopPreviewLiveStream(itemView: SingleColumnItemView) {
        itemView.stopPreviewLiveStream()
        playStreamView.remove(itemView)
    }

    private fun stopAllPreviewLiveStream() {
        playStreamView.forEach { it.stopPreviewLiveStream() }
        playStreamView.clear()
    }

    private fun pauseAllLiveStream() {
        playStreamView.forEach { itemView ->
            if (willEnterRoomView != itemView) {
                itemView.stopPreviewLiveStream()
            }
        }
    }

    fun refreshData() {
        if (isLoading) return

        if (System.currentTimeMillis() - loadingTime < REFRESH_TIME_INTERVAL) {
            removeCallbacks(runnable)
            postDelayed(runnable, REFRESH_TIME_INTERVAL)
            return
        }

        loadingTime = System.currentTimeMillis()
        isLoading = true
        liveListViewPagerAdapter.refreshData {
            swipeRefreshLayout.isRefreshing = false
            post {
                isLoading = false
                if (!isResumed) {
                    LOGGER.info("the current activity is not resumed")
                    return@post
                }
                stopAllPreviewLiveStream()
                val currentPosition = liveListViewPager.currentItem
                val currentView = liveListViewPager.findViewByPosition(currentPosition)
                (currentView as? SingleColumnItemView)?.let {
                    startPreviewLiveStream(it, false)
                }
            }
        }
    }

    private fun onPictureInPictureRoomIdChanged(roomId: String?) {
        if (TextUtils.isEmpty(roomId)) {
            playStreamView.forEach { itemView ->
                if (itemView.isPauseByPictureInPicture()) {
                    itemView.startPreviewLiveStream(false)
                }
            }
        }
    }

    private val lifecycleObserver = LifecycleEventObserver { _, event ->
        when (event) {
            Lifecycle.Event.ON_RESUME -> {
                isResumed = true
                willEnterRoomView = null
                if (isInit) {
                    refreshData()
                } else {
                    isInit = true
                }
            }

            Lifecycle.Event.ON_PAUSE -> {
                isResumed = false
                pauseAllLiveStream()
            }

            else -> {}
        }
    }

    private val runnable = Runnable {
        if (!isLoading) {
            swipeRefreshLayout.isRefreshing = false
        }
    }
}
